### PR TITLE
subs() comparing key hashes

### PR DESCRIPTION
**Pull request by madsbk**
This PR makes `subs()` compare hashes of keys instead of comparing type and equality of tuple items.

This follows the [graph specification](https://docs.dask.org/en/latest/spec.html#definitions): `"A key is any hashable value that is not a task"`, which implies that different typed keys that hashes to the same value are the same key?.

This is motivated by my [work on culling](https://github.com/dask/dask/pull/6510) where keys containing `int` and `numpy.int` are substituted. 

```python
assert hash(('mykey', 42)) == hash(('mykey', np.int64(42)))
```

Finally, I expect this to be slightly faster than the current implementation.

- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

**Comment by mrocklin** [comment]
Thanks @madsbk 

cc'ing @eriknw who might have thoughts on this

**Comment by eriknw** [comment]
I wouldn't rely only on hashes, which seems far too risky.  But you're also right that relying on *types* may be overly restrictive.

How about `len({x, y}) == 1`?  This checks both hash and equality (as one would expect), and should be pretty fast.

**Comment by eriknw** [comment]
Aha, this is even better:
```python
x in {y}
```

**Comment by eriknw** [comment]
> This follows the graph specification: `"A key is any hashable value that is not a task"`, which implies that different typed keys that hashes to the same value are the same key?.

To clarify, "hashable" does not imply only using the hash value.  Typically, it also implies using equality as well, hence why I don't think we should only use hash values.

In general, I think it's *probably* okay to relax the type check.  This definitely needs to go in the release notes, because I wouldn't be surprised if this change breaks some code somewhere (unlikely, but possible).

**Comment by madsbk** [comment]
> To clarify, "hashable" does not imply only using the hash value. Typically, it also implies using equality as well, hence why I don't think we should only use hash values.

Make sense and I really like `x in {y}` !

File: dask/core.py [hunk 1]
```diff
@@ -287,6 +287,7 @@ def subs(task, key, val):
     >>> subs((inc, 'x'), 'x', 1)  # doctest: +SKIP
     (inc, 1)
     """
+    hash_key = {key}
```
Comment (eriknw) [inline dask/core.py:290]: Teeny tiny nit: I would probably move `hash_key` definition down 10 lines or so to right before the `for` loop.  Fine as is though.

**Comment by eriknw** [comment]
LGTM.  I like it!  Thanks @madsbk

File: dask/core.py (hunk 1 above)
Comment (madsbk) [reply dask/core.py]: Fixed

**Comment by madsbk** [comment]
@mrocklin, I think this is ready to be merged. 
I don't know what is going on with CI. It complains about `black reformat` of files I haven't touched. On my machine, both black and flake8 passes.

**Comment by mrocklin** [comment]
Black recently updated.  I think that Tom resolved this on the Dask side with #6568 .  Pushing an empty commit to trigger CI

**Pull request merged**